compiler/ast/types.nim

```diff
@@ -1398,3 +1400,123 @@ proc isObjLackingTypeField*(typ: PType): bool {.inline.} =
+  ## reachable through a ``ref`` type part of `t`. `isInd` indicates whether
+  ## the analysed `t` was reached through following a ``ref`` type
+  if t == nil or tfAcyclic in t.flags:
+    return false
+
+  # if the type is marked as acyclic, it is asserted to never being part of a
+  # reference cycle at run-time, meaning that we don't need to follow it further
+  let t = t.skipTypes(abstractInst - {tyTypeDesc})
+  if tfAcyclic in t.flags:
+    return false
+  # we're not interested in the difference between named and unnamed tuples nor
+  # distinct types, so use ``sameBackendType`` instead of ``sameType``
+  if isInd and sameBackendType(t, search):
+    # the searched for type is reachable from itself through a ref indirection
+    return true
+
+  let trace = getAttachedOp(g, t, attachedTrace)
+  if trace != nil and sfOverriden in trace.flags:
+    # the type has a custom trace hook. We take it as implying that reasoning
+    # about the type is not possible, and have to assume that a cycle through
+    # the type is possible
+    return true
+
+  case t.kind
+  of tyRef:
+    let base = t.base.skipTypes(abstractInst - {tyTypeDesc})
+    if base.kind == tyObject and not isFinal(base):
+      # a polymorphic ``ref``. We have to assume that the dynamic type contains
+      # the `startId` type somewhere
+      result = tfAcyclic notin base.flags
+    else:
+      # a static ``ref``. Remember that we're analysing a type reachable
+      # through a ref indirection and check the element type
+      result = check(marker, g, base, search, isInd=true)
+
+  of tyProc:
+    # since the env type is dynamic (only known at run-time), we don't know if a
+    # cycle to the type we're searching for is possible (or not possible) -- we
+    # have to assume that it's possible
+    result = t.callConv == ccClosure
+  else:
+    result = trav(marker, g, t, search, isInd)
+
+proc travNode(marker: var IntSet, g: ModuleGraph, n: PNode, search: PType,
+              isInd: bool): bool =
+  ## Traverses the fields of the given record `n` and checks if the type
+  ## identified by `startId` is reachable through one of them. Returns the
+  ## result.
+  result = false
+  if n != nil:
+    case n.kind
+    of nkSym:
+      # cursor fields are non-owning -- they can't keep references alive and
+      # thus can't cause a reference cycle
+      if sfCursor notin n.sym.flags:
+        result = check(marker, g, n.typ, search, isInd)
+    of nkWithSons:
+      for it in n.items:
+        result = travNode(marker, g, it, search, isInd)
+        if result: return
+    of nkWithoutSons - {nkSym}:
+      discard "not relevant"
+
+proc trav(marker: var IntSet, g: ModuleGraph, t: PType, search: PType,
+          isInd: bool): bool =
+  let t = t.skipTypes(abstractInst-{tyTypeDesc})
+
+  case t.kind
+  of tyTuple, tySequence, tyArray:
+    # value types that can keep something alive
+    for i in 0..<t.len:
+      result = check(marker, g, t[i], search, isInd)
```

Comment:
Yep, you're right. I accidentally removed the short-circuiting logic there.

Reply:
I used `break` instead of `return` to favour fall through/linear control flow.

```suggestion
      result = check(marker, g, t[i], search, isInd)
      if result:
        break
```